gh-58451: Add optional delete_on_close parameter to NamedTemporaryFile (#97015)
python/cpython: Doc/library/tempfile.rst

      Added *errors* parameter.

.. function:: NamedTemporaryFile(mode='w+b', buffering=-1, encoding=None, newline=None, suffix=None, prefix=None, dir=None, delete=True, *, errors=None, delete_on_close=True)

   This function operates exactly as :func:`TemporaryFile` does, except the
   following differences:

   * The file is guaranteed to have a visible name in the file system

>      (on Unix, the directory entry is not unlinked).

Reviewer: I know this line was here before, but it doesn't make any sense to me. Should we mention this somewhere else?

Author: @eryksun , I think **zooba** talks about the line "(on Unix, the directory entry is not unlinked)." What do you think, does it make scene to move it somewhere else?

Reviewer 2: On POSIX, a file created by `TemoraryFile()` does not have a visible name in the filesystem. The implementation tries to use the `O_TMPFILE` flag if it's supported (Linux 3.11+), which creates an anonymous file. If `O_TMPFILE` isn't supported, the file is created with a unique name, and then it's immediately unlinked while the file descriptor is still open.

On Windows, `TemporaryFile` is aliased to `NamedTemporaryFile`. There isn't a compelling reason for this decision. It's fine to delete a file that's opened with the `O_TEMPORARY` flag. On Windows 10 and 11, if it's on an NTFS filesystem, the file gets unlinked immediately. Prior to Windows 10, or on filesystems other than NTFS, the file is deleted but not unlinked. In the latter state, a file has a visible entry in the directory, but it can't be opened again for any access, and it gets unlinked as soon as the last handle to it is closed.

Author: @eryksun , thanks for explanation. So, would you agree, that the statement _on Unix, the directory entry is not unlinked_ does not really make too much scene the way it is written now? Would it be better to either remove it completely or to put a broader explanation, similar to what you provided?

Reviewer 2: It's in the context of clarifying that `NamedTemporaryFile()` acts "exactly" like `TemporaryFile()`. In that context it makes sense. But I'd prefer that the first sentence and the two bullet points were replaced with something like the following:

```text
   This function returns a file that is guaranteed to have a visible name in
   the file system. To manage the named file, it extends the parameters of
   :func:`TemporaryFile` with *delete* and *delete_on_close* parameters that
   determine whether and how the named file should be automatically deleted.
```

Author: I have now placed the statement about unlinking on Unix to a different place and slightly reworked it to make a reference to TemporaryFile(), without which I agree it was a bit confusing.

I have also implemented rewording from @eryksun

   * There is more granularity in the deletion behaviour of the file
     (see *delete_on_close* below)

   The returned object is always a :term:`file-like object` whose :attr:`!file`
   attribute is the underlying true file object. This :term:`file-like object` can be
   used in a :keyword:`with` statement, just like a normal file.  The name of the
   temporary file can be retrieved from the :attr:`name` attribute of the
   returned file-like object.

   If *delete* is true (the default) and *delete_on_close* is true (the
   default), the file is deleted as soon as it is closed. If *delete* is true
   and *delete_on_close* is false, the file is deleted on context manager exit

>    only, if no context manager was used, then the file is deleted only when the
>    :term:`file-like object` is finalized and hence deletion is not always
>    guaranteed in this case (see :meth:`object.__del__`). If *delete* is false,
>    the value of *delete_on_close* is ignored.

Reviewer: ```suggestion
   only, or else when the :term:`file-like object` is finalized. Deletion is not
   always guaranteed in this case (see :meth:`object.__del__`). If *delete* is
   false, the value of *delete_on_close* is ignored.
```

Author: I have now implemented this locally on my PC.

   While the named temporary file is open, the file can always be opened again
   on POSIX. On Windows, it can be opened again if *delete* is false, or if
   *delete_on_close* is false, or if the additional open shares delete access
   (e.g. by calling :func:`os.open` with the flag ``O_TEMPORARY``).  On
   Windows, if *delete* is true and *delete_on_close* is false, additional
   opens that do not share delete access (e.g. via builtin :func:`open`) must
   be closed before exiting the context manager, else the :func:`os.unlink`

>    call on context manager exit will fail with a :exc:`PermissionError`.

Reviewer: Can we reframe this as:

* the file can be opened again using its name
* on Windows, subsequent opens should share delete access (e.g.)
* however, when `delete` or `delete_on_close` are false, sharing delete access is not necessary, but you should ensure the file is closed before leaving the context manager

This sets it up more positively for the user - what _should_ they do, as opposed to what could go wrong.

Author: Please check how I have written this in the new version. Not sure everybody would agree, but I have now written this information the way I would want it to be explained for me without the risk of twisting my brains.

**Additional question:** does it make sense to  write unit tests to cover all these combinations of opening the file the second time, while it is already open.  I have tested this for myself (on Windows) to make sure it is all correct, but I am not sure that it shall go to unit tests.

   To use the name of the temporary file to open the closed file second time,
   either make sure not to delete the file upon closure (set the *delete*
   parameter to be false) or, in case the temporary file is created in a
   :keyword:`with` statement, set the *delete_on_close* parameter to be false.
   The latter approach is recommended as it provides assistance in automatic

>    cleaning of the temporary file upon the context manager exit.

Reviewer: This paragraph seems redundant with the previous one?

Author: The previous one talks about opening the file, while it is still open, This section talks about opening the file after it has already been closed, which is a bit different case. I agree one can work it out from the information at the beginning of the section, but I thought that because it represents arguably the most frequent use, then may be it would not harm to have this paragraph.

Any way, I have now moved it to a slightly different place and have re-written to make sure a reader understands, that this is not a new information, but the practical conclusion from the previous paragraph.

      Added *errors* parameter.

.. function:: NamedTemporaryFile(mode='w+b', buffering=-1, encoding=None, newline=None, suffix=None, prefix=None, dir=None, delete=True, *, errors=None, delete_on_close=True)

   This function operates exactly as :func:`TemporaryFile` does, except the
   following differences:

   * The file is guaranteed to have a visible name in the file system (on Unix,
     the directory entry is not unlinked).

   * There is more granularity in the deletion behaviour of the file (see

>      ``delete_on_close`` below)

Reviewer 3: ```suggestion
   * The file is guaranteed to have a visible name in the file system
     (on Unix, the directory entry is not unlinked).
   * There is more granularity in the deletion behaviour of the file
     (see *delete_on_close* below)
```
Fixed *delete_on_close* markup, rewrapped, removed newline.

     the file system.
   * To manage the named file, it extends the parameters of
     :func:`TemporaryFile` with *delete* and *delete_on_close* parameters that
     determine whether and how the named file should be automatically deleted.

   The returned object is always a :term:`file-like object` whose :attr:`!file`
   attribute is the underlying true file object. This :term:`file-like object`
   can be used in a :keyword:`with` statement, just like a normal file.  The
   name of the temporary file can be retrieved from the :attr:`name` attribute
   of the returned file-like object. On Unix, unlike with the
   :func:`TemporaryFile`, the directory entry does not get unlinked immediately
   after the file creation.

   If *delete* is true (the default) and *delete_on_close* is true (the
   default), the file is deleted as soon as it is closed. If *delete* is true
   and *delete_on_close* is false, the file is deleted on context manager exit
   only, or else when the :term:`file-like object` is finalized. Deletion is not
   always guaranteed in this case (see :meth:`object.__del__`). If *delete* is
   false, the value of *delete_on_close* is ignored.

>    Therefore to use the name of the temporary file to open the closed file
>    second time, either make sure not to delete the file upon closure (set the

Reviewer: ```suggestion
   Therefore to use the name of the temporary file to reopen the file after
   closing it, either make sure not to delete the file upon closure (set the
```

   *delete* parameter to be false) or, in case the temporary file is created in
   a :keyword:`with` statement, set the *delete_on_close* parameter to be false.
   The latter approach is recommended as it provides assistance in automatic
   cleaning of the temporary file upon the context manager exit.

   Opening the temporary file again by its name while it is still open works as
   follows:

   * On POSIX the file can always be opened again.

>    * On Windows to open the file again while it is open make sure that at least
>      one of the following conditions are fulfilled:

Reviewer: ```suggestion
   * On Windows, make sure that at least one of the following conditions are
     fulfilled:
```

    >>>
    # file is now closed and removed

    # create a temporary file using a context manager, note the name,
    # close the file, use the name to open the file again
    >>> with tempfile.TemporaryFile(delete_on_close=False) as fp:
    ...     fp.write(b'Hello world!')
    ...     fp_name = fp.name
    ...     fp.close()
    # the file is closed, but not removed
    # open the file again by using its name

>     ...     f = open(fp_name)

Reviewer: `fp.name` is still valid here, right? Why capture it?

Reviewer: Also, maybe use this one in its own nested `with` block?

Author: I captured the `fp.name` to make more explicit, but I agree, it is redundant, so I removed it now.
Also nested `with` block is added.

>     ...     f.seek(0)

Reviewer: We shouldn't need to seek, right?

Author: Correct, I removed it now
